kernel/src/checkpoint/log_replay.rs

```diff
@@ -285,147 +285,145 @@ impl CheckpointVisitor<'_> {
+                let last_updated: i64 = last_updated;
+                if last_updated <= retention_ts {
                     // Transaction is old, exclude it from checkpoint
-                    return Ok(false);
+                    return Ok(Some(false));
                 }
             }
             // Note: transactions without last_updated are kept for backward compatibility
         }
 
         // If the app ID already exists in the set, the insertion will return false,
         // indicating that this is a duplicate.
-        if !self.seen_txns.insert(app_id.to_string()) {
-            return Ok(false);
-        }
-
-        // Valid, non-duplicate txn action to be included
-        Ok(true)
+        Ok(Some(self.seen_txns.insert(app_id.to_string())))
     }
 
     /// Determines if a row in the batch should be included in the checkpoint.
     ///
-    /// This method checks each action type in sequence, short-circuiting as soon as a valid action is found.
+    /// This method checks each action type in sequence, short-circuiting when:
+    /// - A valid action is found (`Some(true)`)
+    /// - A suppressed action is found (`Some(false)`)
+    /// - An error occurs (propagated immediately)
+    ///
     /// Actions are checked in order of expected frequency of occurrence to optimize performance:
     /// 1. File actions (most frequent)
     /// 2. Txn actions
     /// 3. Protocol & Metadata actions (least frequent)
     ///
-    /// Returns Ok(true) if the row should be included in the checkpoint.
-    /// Returns Ok(false) if the row should be skipped.
-    /// Returns Err(...) if any validation or extraction failed.
+    /// Returns `Ok(true)` if the row should be included in the checkpoint.
+    /// Returns `Ok(false)` if the row should be skipped.
+    /// Returns `Err(...)` if any validation or extraction failed.
     pub(crate) fn is_valid_action<'a>(
         &mut self,
         i: usize,
         getters: &[&'a dyn GetData<'a>],
     ) -> DeltaResult<bool> {
-        // The `||` operator short-circuits the evaluation, so if any of the checks return true,
-        // the rest will not be evaluated.
-        let is_valid = self.check_file_action(i, getters)?
-            || self.check_txn_action(i, getters)?
-            || self.check_protocol_action(i, getters[10])?
-            || self.check_metadata_action(i, getters[9])?;
+        // Check each action type in sequence, short-circuiting when an action is found
+        let is_valid = if let Some(result) = self.check_file_action(i, getters)? {
+            result
+        } else if let Some(result) = self.check_txn_action(i, getters)? {
+            result
+        } else if let Some(result) = self.check_protocol_action(i, getters[10])? {
+            result
+        } else if let Some(result) = self.check_metadata_action(i, getters[9])? {
+            result
+        } else {
+            false // If all returned None, no valid action found
+        };
```

Comment:
per clippy suggestion above:
```suggestion
        } else {
            self.check_metadata_action(i, getters[9])?.unwrap_or(false)
        };
```
(tho it may insist on replacing `unwrap_or(false)` with `unwrap_or_default()`)